Add the guest effective CPU and memory to kubevirt_vm_resource_requests metric

sradco opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
#### Before this PR:
The kubevirt_vm_resource_requests metric didnt report the effective values as expected.

#### After this PR:
This PR updates the kubevirt_vm_resource_requests metrics
to report  additional "source"="guest_effective" and updated
its value to the effective value of the guest.

Example:
For a  VM name vm1 in namespace ns1 with:
- Effective memory: 2Gi (2147483648 bytes)
- Effective vCPUs: 4 (e.g., 2 cores × 2 threads × 1 socket)
- Threads and sockets are reported as 1 for default-only compatibility

Output:
```
kubevirt_vm_resource_requests{name="vm1", namespace="ns1", resource="memory", unit="bytes", source="guest_effective"} 2147483648

kubevirt_vm_resource_requests{name="vm1", namespace="ns1", resource="cpu", unit="cores", source="guest_effective"} 4
```

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
- Fixes # https://issues.redhat.com/browse/CNV-61831
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Report allocated CPU and memory requests as simplified metrics with source="guest_effective" label , showing final values after applying instance types, preferences, and hierarchy.```

sradco commented:
@machadovilaca @avlitman @nunnatsa please review this enhancement proposal.

iholder101 commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -480,6 +522,38 @@ func collectCpuResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []ope
+		cr = append(cr, operatormetrics.CollectorResult{
+			Metric: vmResourceRequests,
+			Value:  float64(allocatedVCPUs),
+			Labels: []string{vm.Name, vm.Namespace, "cpu", "cores", "allocated"},
```
The phrasing of "allocated" is a bit confusing IMO.
Can we change this to something like "virt-launcher resources", to ensure we make it clear the resources are not allocated to the guest itself.

sradco replied:
Updated to "guest_effective"

sradco commented:
@machadovilaca @enp0s3 @avlitman  Please review.

machadovilaca commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -480,6 +522,38 @@ func collectCpuResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []ope
+	}
+
+	// Also report allocated CPU limits if they exist
+	if cpuLimits := vm.Spec.Template.Spec.Domain.Resources.Limits.Cpu(); !cpuLimits.IsZero() {
```
this seems to be the same as line 510

sradco replied:
Removed

machadovilaca commented on pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:
```diff
@@ -346,7 +368,29 @@ var _ = Describe("VM Stats Collector", func() {
 
-			expectDefaultCPUResourceRequests(crs[2:])
+			// Find guest_effective memory requests in the results
+			var allocatedRequestsFound bool
+			for _, cr := range crs {
+				if len(cr.Labels) == 5 && cr.Labels[4] == "guest_effective" && cr.Labels[2] == "memory" {
+					if cr.Metric.GetOpts().Name == "kubevirt_vm_resource_requests" {
+						By("checking the guest_effective memory requests")
+						Expect(cr.Value).To(BeEquivalentTo(1024))
+						Expect(cr.Labels).To(Equal([]string{"testvm", "test-ns", "memory", "bytes", "guest_effective"}))
+						allocatedRequestsFound = true
+						break
+					}
+				}
+			}
+			Expect(allocatedRequestsFound).To(BeTrue(), "Should find guest_effective memory requests")
+
+			// Find and validate CPU default metrics
+			var cpuDefaultMetrics []operatormetrics.CollectorResult
+			for _, cr := range crs {
+				if len(cr.Labels) == 5 && cr.Labels[4] == "default" && cr.Labels[2] == "cpu" {
+					cpuDefaultMetrics = append(cpuDefaultMetrics, cr)
+				}
+			}
+			expectDefaultCPUResourceRequests(cpuDefaultMetrics)
```
can we skip all this logic here, since we already have test below for the guest effective?
if not can we move it to a function

and the verification has a lot of identation and uses a lot of labels indexes, if you can hide them in functions like "has guest effective memory resource request"

sradco replied:
Updated

sradco commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -389,6 +395,42 @@ func collectMemoryResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []
+		cr = append(cr, operatormetrics.CollectorResult{
+			Metric: vmResourceRequests,
+			Value:  float64(allocatedMemory.Value()),
+			Labels: []string{vm.Name, vm.Namespace, "memory", "bytes", "guest_effective"},
```
I was asked to update the name, to reflect that this is the actual **guest** requested memory. 
If there are several places where the memory is defined , this should reflect the end effective memory.

sradco commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -389,6 +395,42 @@ func collectMemoryResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []
+	}
+
+	// Also report allocated memory limits if they exist
+	if memoryLimits := vm.Spec.Template.Spec.Domain.Resources.Limits.Memory(); !memoryLimits.IsZero() {
```
You are correct, for limits we already report the effective memory and cpu. Removing.

sradco replied:
Removed

sradco commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -480,6 +522,38 @@ func collectCpuResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []ope
+			Value:  float64(allocatedVCPUs),
+			Labels: []string{vm.Name, vm.Namespace, "cpu", "cores", "guest_effective"},
+		})
+	}
```
Unfortunately we have dashboards that depend on the existing values and I know there are others that are already using them, so we can't deprecate them easily.

sradco commented:
/hold

sradco commented:
/test pull-kubevirt-e2e-k8s-1.34-sig-network

sradco committed:
Update CPU and Memory requested metric

This PR updates the kubevirt_vm_resource_requests
metric and adds source for the effective value of the guest.

Signed-off-by: Shirly Radco <sradco@redhat.com>

sradco commented:
/test pull-kubevirt-check-unassigned-tests
/test pull-kubevirt-e2e-k8s-1.34-sig-compute-serial

enp0s3 commented on pkg/monitoring/metrics/virt-controller/vmstats_collector.go:
```diff
@@ -403,10 +491,30 @@ func collectMemoryResourceLimitsFromDomainResources(vm *k6tv1.VirtualMachine) []
+	}
+
+	vmi := &k6tv1.VirtualMachineInstance{Spec: vm.Spec.Template.Spec}
+	allocatedMemory := vcpu.GetVirtualMemory(vmi)
```
@sradco You can reuse create `collectMemoryResourceRequestsFromDomainResources` and create the `guest_effective` label there:
```
	effectiveMemory := guestMemory
	if effectiveMemory == nil || effectiveMemory.IsZero() {
		effectiveMemory = memoryRequested
	}

	cr = append(cr, operatormetrics.CollectorResult{
		Metric: vmResourceRequests,
		Value:  float64(effectiveMemory.Value()),
		Labels: []string{vm.Name, vm.Namespace, "memory", "bytes", "guest_effective"},
	})
```

sradco replied:
`collectMemoryResourceRequestsFromDomainResources` reports the raw configured fields as separate metrics: `Resources.Requests.memory` ("domain"), `Memory.Guest` ("guest"), and `Memory.Hugepages.PageSize` ("hugepages").
It does not apply hierarchy or fallbacks and it doesn’t consider limits when requests are unset.
It’s essentially “what is set where,” not “what will be used.”
`vcpu.GetVirtualMemory` returns a single effective guest memory value after applying KubeVirt’s hierarchy.

enp0s3 commented on pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:
```diff
@@ -1033,3 +1190,29 @@ func expectVMs(
 	expectVM(filterResultsByVM(results, name+"-2", namespace), name+"-2", namespace)
 }
+
+func expectDefaultCPUResourceRequests(crs []operatormetrics.CollectorResult) {
```
@sradco We prefer no to use assertions in helper functions, can we please refactor this helper function to return the default CPU requests and an error, and the test itself to assert on the returned error and on the returned defaults?

avlitman commented:
/approve

enp0s3 reviewed (commented):
/lgtm
